<a href="https://colab.research.google.com/github/ninja-josh/sales-predictions/blob/main/sales_predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predicting Sales

In [10]:
import pandas as pd

In [11]:
url = 'https://raw.githubusercontent.com/ninja-josh/sales-predictions/main/data/sales_predictions.csv?token=AVQW6F6BCQDXFXMKNJK65HLBN4XTA'
df = pd.read_csv(url)
df.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

# Data Exploration

In [12]:
# Number of columns and rows
df.shape

(8523, 12)

In [13]:
# check datatypes

df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [14]:
df.duplicated().sum()

0

# Missing Data

In [15]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

columns 'Item_Weight' and 'Outlet_Size' are missing data, let's take a closer look.

In [17]:
df.describe()

Item_Weight  ...  Item_Outlet_Sales
count  7060.000000  ...        8523.000000
mean     12.857645  ...        2181.288914
std       4.643456  ...        1706.499616
min       4.555000  ...          33.290000
25%       8.773750  ...         834.247400
50%      12.600000  ...        1794.331000
75%      16.850000  ...        3101.296400
max      21.350000  ...       13086.964800

[8 rows x 5 columns]

In [16]:
df[df['Item_Weight'].isna()].describe()

Item_Weight  ...  Item_Outlet_Sales
count          0.0  ...        1463.000000
mean           NaN  ...        2483.677474
std            NaN  ...        2347.516621
min            NaN  ...          33.955800
25%            NaN  ...         399.480000
50%            NaN  ...        1845.597600
75%            NaN  ...        3968.168000
max            NaN  ...       13086.964800

[8 rows x 5 columns]

In [19]:
df[df['Outlet_Establishment_Year'] == 1985].describe()

Item_Weight  ...  Item_Outlet_Sales
count          0.0  ...        1463.000000
mean           NaN  ...        2483.677474
std            NaN  ...        2347.516621
min            NaN  ...          33.955800
25%            NaN  ...         399.480000
50%            NaN  ...        1845.597600
75%            NaN  ...        3968.168000
max            NaN  ...       13086.964800

[8 rows x 5 columns]

rows missing data in 'Item_Weight' are all from 1985.  All data from 1985 is missing 'Item_Weight' data.  This is the earliest year of the data, so I don't feel we will lose too much data by dropping them.

In [21]:
df = df.dropna(subset=['Item_Weight'])
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [18]:
df[df['Outlet_Size'].isna()].describe()

Item_Weight  ...  Item_Outlet_Sales
count  2410.000000  ...        2410.000000
mean     12.778479  ...        1822.626947
std       4.589825  ...        1557.858015
min       4.590000  ...          33.290000
25%       8.785000  ...         554.777850
50%      12.600000  ...        1443.454400
75%      16.700000  ...        2681.509500
max      21.350000  ...        9664.752800

[8 rows x 5 columns]

I'm not seeing a pattern, what if we include all columns?

In [22]:
df[df['Outlet_Size'].isna()].describe(include='all')

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
count             2410  2410.000000  ...               2410        2410.000000
unique            1393          NaN  ...                  2                NaN
top              FDR23          NaN  ...  Supermarket Type1                NaN
freq                 3          NaN  ...               1855                NaN
mean               NaN    12.778479  ...                NaN        1822.626947
std                NaN     4.589825  ...                NaN        1557.858015
min                NaN     4.590000  ...                NaN          33.290000
25%                NaN     8.785000  ...                NaN         554.777850
50%                NaN    12.600000  ...                NaN        1443.454400
75%                NaN    16.700000  ...                NaN        2681.509500
max                NaN    21.350000  ...                NaN        9664.752800

[11 rows x 12 columns]

I notice that 1855 of the 2410 observations that are missing data for 'Outlet_Size' are 'Outlet_Location_Type' == Tier 2.  Let's see if there is any pattern in rows of Tier 2 location type.

In [25]:
df[df['Outlet_Location_Type'] == 'Tier 2'].describe(include='all')

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
count             2785  2785.000000  ...               2785        2785.000000
unique            1454          NaN  ...                  1                NaN
top              FDV60          NaN  ...  Supermarket Type1                NaN
freq                 3          NaN  ...               2785                NaN
mean               NaN    12.768628  ...                NaN        2323.990559
std                NaN     4.607141  ...                NaN        1520.543543
min                NaN     4.555000  ...                NaN          99.870000
25%                NaN     8.750000  ...                NaN        1171.808000
50%                NaN    12.500000  ...                NaN        2004.058000
75%                NaN    16.750000  ...                NaN        3110.617600
max                NaN    21.350000  ...                NaN        9664.752800

[11 rows x 12 columns]

I see that all of the rows where 'Outlet_Location_type' is Tier 2 are also 'Outlet_Size' == Small.  Since a large majority of the missing data has 'Outlet_Location_Type' == Tier 2, and all Tier 2 rows have Outlet_Size == Small, we can conclude that most of the missing data should be of 'Outlet_Size' == Small.

In [27]:
df['Outlet_Size'].fillna('Small', inplace=True)
df.isna().sum()

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

# Unneeded Columns

In [29]:
df.describe(include='all')

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
count             7060  7060.000000  ...               7060        7060.000000
unique            1555          NaN  ...                  3                NaN
top              FDF22          NaN  ...  Supermarket Type1                NaN
freq                 8          NaN  ...               5577                NaN
mean               NaN    12.857645  ...                NaN        2118.626808
std                NaN     4.643456  ...                NaN        1533.445842
min                NaN     4.555000  ...                NaN          33.290000
25%                NaN     8.773750  ...                NaN         922.798800
50%                NaN    12.600000  ...                NaN        1789.670400
75%                NaN    16.850000  ...                NaN        2966.139000
max                NaN    21.350000  ...                NaN       10256.649000

[11 rows x 12 columns]

'Item_Identifier' sounds like an an identification columns, the values look like labels, and thre are 1555 different values.  I'm going to drop this column because I think it is not carrying useful informaiton

In [30]:
df = df.drop(columns = 'Item_Identifier')

# Inconsistent Categories

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7060 entries, 0 to 8522
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Weight                7060 non-null   float64
 1   Item_Fat_Content           7060 non-null   object 
 2   Item_Visibility            7060 non-null   float64
 3   Item_Type                  7060 non-null   object 
 4   Item_MRP                   7060 non-null   float64
 5   Outlet_Identifier          7060 non-null   object 
 6   Outlet_Establishment_Year  7060 non-null   int64  
 7   Outlet_Size                7060 non-null   object 
 8   Outlet_Location_Type       7060 non-null   object 
 9   Outlet_Type                7060 non-null   object 
 10  Item_Outlet_Sales          7060 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 661.9+ KB


In [34]:
categorical_columns = ['Item_Fat_Content', 'Item_Type','Outlet_Identifier','Outlet_Size','Outlet_Location_Type','Outlet_Type']

for col in categorical_columns:
  print('\n', col)
  print(df[col].value_counts())



 Item_Fat_Content
Low Fat    4222
Regular    2388
LF          260
reg         106
low fat      84
Name: Item_Fat_Content, dtype: int64

 Item_Type
Fruits and Vegetables    1019
Snack Foods               988
Household                 759
Frozen Foods              718
Dairy                     566
Canned                    539
Baking Goods              536
Health and Hygiene        430
Soft Drinks               374
Meat                      337
Breads                    204
Hard Drinks               183
Others                    137
Starchy Foods             130
Breakfast                  89
Seafood                    51
Name: Item_Type, dtype: int64

 Outlet_Identifier
OUT013    932
OUT035    930
OUT046    930
OUT049    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64

 Outlet_Size
Small     4270
Medium    1858
High       932
Name: Outlet_Size, dtype: int64

 Outlet_Location_Type
Tier 2    2785
Tier 3    2415
Tier 1    1860
Name: Outlet_

Item Fat Content seems to have some redundant values.  Let's normalize those.

In [36]:
df['Item_Fat_Content'].replace({'LF':'Low Fat',
                                'low fat': 'Low Fat',
                                'reg': 'Regular'},
                               inplace=True)
df['Item_Fat_Content'].value_counts()

Low Fat    4566
Regular    2494
Name: Item_Fat_Content, dtype: int64

In [37]:
df.describe()

Item_Weight  ...  Item_Outlet_Sales
count  7060.000000  ...        7060.000000
mean     12.857645  ...        2118.626808
std       4.643456  ...        1533.445842
min       4.555000  ...          33.290000
25%       8.773750  ...         922.798800
50%      12.600000  ...        1789.670400
75%      16.850000  ...        2966.139000
max      21.350000  ...       10256.649000

[8 rows x 5 columns]